faster aob

juzt3 opened the pull request:
implementation of the Boyer Moore algorithm

in my tests it went from 7.5 secs using aob1 to 4.5 secs using boyerMoore in my specific game and pattern
it means a gain of around 40%

juzt3 committed:
faster aob

implementation of the Boyer Moore algorithm

qb-0 commented:
Interesting. `aob1` was actually @IagoBeuller's attempt to realize the boyer moore algorithm. If that's a better approach we could prob replace `aob1` with your code.

@IagoBeuller What do you think?

juzt3 committed:
enum memory regions of process

i finished this implementation becose i needed it

juzt3 committed:
aob scan through memory pages added

with boyer moore and aobscan through memory pages directly with nim is even faster now

juzt3 commented:
i have added aob scan through memory pages and the time needed to find my pattern in my game is 3.4 secs now

IagoBeuller commented:
> Interesting. `aob1` was actually @IagoBeuller's attempt to realize the boyer moore algorithm. If that's a better approach we could prob replace `aob1` with your code.
> 
> @IagoBeuller What do you think?

Of course, if the algorithm is faster, but I didn't test it yet. I also did a function to find patterns scanning page by page. I would make some improvements to the isMatch proc btw.

juzt3 committed:
clean up

juzt3 committed:
NtReadVirtualMemory added

this is arround 5% faster than ReadProcessMemory
https://www.unknowncheats.me/forum/general-programming-and-reversing/230813-readprocessmemory-vs-ntreadvirtualmemory-performance-benchmark-comparison.html

juzt3 committed:
Update memcore.nim

juzt3 commented:
i ran out of ideas. do you think is possible to get better perfomance?

IagoBeuller commented:
> i ran out of ideas. do you think is possible to get better perfomance?

The algorithm is accessing and getting values from a position in the sequence/array even when its not needed because of "??". This costs performance. To solve, I did this way:

```nim
proc isMatch(buffer: seq[byte], pattern: seq[int], start: int): bool =
  for i in 0..<pattern.len:
    let p = pattern[i]
    if p == doubleWildCardInt:
      continue
    
    let b = buffer[start + i].int
    if p == wildCardIntL:
      if b.toHex(1)[0] != pattern[i].toHex(1)[1]:
        return false
    elif p == wildCardIntR:
      if b.toHex(2)[0] != pattern[i].toHex(2)[0]:
        return false
    elif p != b:
      return false
  return true
```

juzt3 commented:
> > i ran out of ideas. do you think is possible to get better perfomance?
> 
> The algorithm is accessing and getting values from a position in the sequence/array even when its not needed because of "??". This costs performance, and how much more "??" the pattern has, the more relevant is the cost. To solve, I did this way:
> 
> ```nim
> proc isMatch(buffer: seq[byte], pattern: seq[int], start: int): bool =
>   for i in 0..<pattern.len:
>     let p = pattern[i]
>     if p == doubleWildCardInt:
>       continue
>     
>     let b = buffer[start + i].int
>     if p == wildCardIntL:
>       if b.toHex(1)[0] != pattern[i].toHex(1)[1]:
>         return false
>     elif p == wildCardIntR:
>       if b.toHex(2)[0] != pattern[i].toHex(2)[0]:
>         return false
>     elif p != b:
>       return false
>   return true
> ```

Great, i tested it and now the time to find my patter in my game is around 2.5 secs. in total a gain of 66% from previous implementation

juzt3 committed:
improvement by IagoBeuller

only read from the array if not a double wild card: ??

juzt3 commented:
i belive that now our aob is faster than ce's imlpementation

juzt3 committed:
true aob algorithm choice
